app/src/main/java/com/example/cactusnotes/signup/SignUpActivity.kt

```diff
@@ -0,0 +1,141 @@
+                sendRegisterRequest()
+            }
+
+        }
+        binding.buttonAccount.setOnClickListener {
+            val intent = Intent(this, LogInActivity::class.java)
+            startActivity(intent)
+        }
+
+    }
+
+    private fun validate(textInputLayout: TextInputLayout): Boolean {
+        val validator = textInputLayout.validator()
+        val field = textInputLayout.editText!!.text.toString()
+        val error = validator.validate(field)
+
+        return if (error == null) {
+            textInputLayout.error = null
+            true
+        } else {
+            textInputLayout.error = getString(error)
+            false
+        }
+
+    }
+
+    private fun TextInputLayout.validator() = when (this) {
+        binding.emailTextInputLayout -> EmailValidator()
+        binding.usernameTextInputLayout -> UsernameValidator()
+        binding.passwordTextInputLayout -> PasswordValidator()
+        else -> throw IllegalArgumentException("No validators are specified for the given TextInputLayout")
+    }
+
+    private fun sendRegisterRequest() {
+
+        val request = RegisterRequest(
+            binding.emailTextInputLayout.editText!!.text.toString(),
+            binding.usernameTextInputLayout.editText!!.text.toString(),
+            binding.passwordTextInputLayout.editText!!.text.toString()
+        )
+        // I don't understand anything at this part
+        // Q1 What doest the enqueue ?
+        api.register(request).enqueue(object : Callback<RegisterResponse> {
+            override fun onResponse(
+                call: Call<RegisterResponse>,
+                response: Response<RegisterResponse>
+            ) {
+                when (response.code()) {
+                    in 200..299 -> registerSuccess()
+                    in 400..499 -> clientSideError(response)
+                    in 500..599 -> serverSideError()
+                    else -> Log.e("SignupActivity", "Unexpected error code")
+                }
+            }
+
+            private fun registerSuccess() {
+                Snackbar.make(
+                    binding.buttonSignUp,
+                    R.string.registered,
+                    Snackbar.LENGTH_LONG
+                ).show()
+
+            }
+
+            private fun serverSideError() {
+                Snackbar.make(
+                    binding.buttonSignUp,
+                    R.string.some_error_occurred,
+                    Snackbar.LENGTH_LONG
+                ).show()
+            }
+
+            // I don't understand anything at this part
+            private fun clientSideError(response: Response<RegisterResponse>) {
+                val responseObj = JSONObject(response.errorBody()!!.string())
+
+                val clientSideErrorMessage = responseObj.getJSONArray("message")
+                    .getJSONObject(0)
+                    .getJSONArray("messages")
+                    .getJSONObject(0)
+                    .getString("message")
```

Comment:
Move these functions out of callback object. They better be in the member level of SignUpActivity class.